core/src/main/java/com/datastax/oss/driver/internal/core/metadata/MetadataManager.java

```diff
@@ -396,33 +423,55 @@ private void acceptSchemaRequest(CompletableFuture<Metadata> future, boolean flu
     }
 
     // The debouncer has flushed, start the actual work.
-    private void startSchemaRequest(CompletableFuture<Metadata> future) {
+    private void startSchemaRequest(CompletableFuture<RefreshSchemaResult> refreshFuture) {
       assert adminExecutor.inEventLoop();
       if (closeWasCalled) {
-        future.complete(metadata);
+        refreshFuture.complete(new RefreshSchemaResult(metadata));
         return;
       }
       if (currentSchemaRefresh == null) {
-        currentSchemaRefresh = future;
+        currentSchemaRefresh = refreshFuture;
         LOG.debug("[{}] Starting schema refresh", logPrefix);
         initControlConnectionForSchema()
             .thenCompose(v -> context.getTopologyMonitor().checkSchemaAgreement())
-            // 1. Query system tables
-            .thenCompose(b -> schemaQueriesFactory.newInstance(future).execute())
-            // 2. Parse the rows into metadata objects, put them in a MetadataRefresh
-            // 3. Apply the MetadataRefresh
-            .thenApplyAsync(this::parseAndApplySchemaRows, adminExecutor)
             .whenComplete(
-                (v, error) -> {
-                  if (error != null) {
-                    currentSchemaRefresh.completeExceptionally(error);
+                (schemaInAgreement, agreementError) -> {
+                  if (agreementError != null) {
+                    refreshFuture.completeExceptionally(agreementError);
+                  } else {
+                    schemaQueriesFactory
+                        .newInstance(
+                            // TODO remove this unused parameter (see JAVA-2582)
+                            null)
```

Comment:
I could have addressed it directly, but it would introduce additional merge conflicts in the OSS/DSE merge. So I created a follow-up ticket, it's not urgent.